### DISCUSSION TITLE
General guide (I use) with this hack

### DISCUSSION THREAD

**Discussion by sideup66** [gtxaspec/wz_mini_hacks#95] [General]
Hi folks,
Little background: I setup this camera hack to use with my wyzecam v3. I do use the stock firmware as it is, but found some frustrating limitations with it. Namely, scheduled reboots and some stability concerns. I have had scenarios in the past where for whatever reason, the cam decides to just spam the router with new connection attempts (usually to a DNS server) to the point that the network tabl will actually fill and crash the entire network. I have setup my own with a scheduled reboot via cron and made a script routerside to monitor active connections from the camera. If it goes over a certain threshold the router will ssh into the camera and issue a reboot command. Prior to this hack I wold rely on wyze cloud and IFTT for a reboot. Don't like this scheme as the wyze apis, amazon, and iftt could have a problem and the command may be never delivered. 

One ask with this is that if someone has more info on how to detect when the camera is actually flooding the network, Id be open to it, as Id like preferably to have the flood detection on the camera itself rather than hoping my router sees the camera misbehaving and acting on it quick enough. As it sits I did catch it in the act of one and while the shell became laggy, I noticed it would remain up. See below for how I setup my stability measures. My hope is that this will help others who find this hack have a simple way to put their device into a stable state. 

1. Setup cron. After installing this on your sd and configuring ssh, I used the below steps to configure crond and have it autostart:
1. i setup a folder for cron in: /opt/wz_mini/etc/crontab
2. in here I have a file called root. This will be important as the crontab will not start.
Inside of here I have my reboot command in the following format:
<code>45 05 * * * /sbin/reboot </code>
note, this time is 1:45 am eastern. Adjust for your local time and needs for reboot. Note the systemtime of the camera is UTC.
Next, I have a cron startup script. This is handy if you need to restart cron (maybe updated the crontab file or added a new job). Code is below of that
<code> 
#!/bin/sh
sleep 15
/opt/wz_mini/tmp/.bin/crond -b -c /opt/wz_mini/etc/crontab/
</code>
This can then be placed in your 
wz_mini.conf file to autostart cron. Id reccomend placing it in the Scripting section, then reboot. Test if cron is running with ps ax | grep crond. If the process is running, you now have a root crontab with a scheduled automatic restart that does NOT rely on any outside service to perform. Its a fair way to keep your device from going stale. 

Next would be setting up the monitoring script. For this I am using DD-WRT firmware and can monitor the connections from the camera every 10 seconds. Once the threshold trips the script simply ssh's the camera and issues a reboot, waits 60 seconds for things to clear, then resumes monitoring. 
https://gist.github.com/sideup66/063ad1aecd65c1d3363afa53d586b50b
Hopefully this helps people out. Thanks!

**Comment by gtxaspec** [comment]
what firmware version do you run on the camera that floods dns?

**Comment by sideup66** [comment]
To note the problem with the network flooding actually was occurring from the camera on stock firmware, the hack did not cause the issue. I was using the hack to take better control of the issue and do some things to improve overall stability (cron reboots and ssh reboots namely) Sorry if that caused any confusion.

At this time the camera is running 4.36.9.139, but I have noticed it on previous builds of the firmware as well.

**Comment by gtxaspec** [comment]
I don't think I've ever looked info the DNS flooding issue, though I have read about it before.  How often does it happen?  Would you be able to get some logs from the camera when it occurs?

**Comment by sideup66** [comment]
It doesnt happen super often, It seems to be pretty sporadic and intermittent. Be happy to pull some logs from the camera the next time It acts up. Would the logs on the camera persist possible reboots? 

Also its a very intermittent issue in my experience, I have not been able to pin down an exact trigger, however I suspect it may have to do wit the camera or the stock firmware thinking there is no connection to the internet and sending tons of new connection requests to what look like dns servers. I can tell this as the connections on the router point to a dns server hosted by my cellphone carrier and all go into a UNREPLIED state. 

Are there applogs anywhere that are specific to the firmware I could find? The standard syslogs I know of in /opt/wz_mini//log didnt seem to contain anything and I couldnt find a CPU or memory contention during my one observed event. Only thing I can reliably point to that may or may not indicate it is when I have the camera running and deny it access to its DNS servers (did this through my pihole) the ssh session on the camera itself can become a bit laggy, however I was never dropped. Maybe something could be seen there?

**Comment by gtxaspec** [comment]
we typically don't log the syslog, so you could add a script that runs on boot that contains:

`logread -F > /opt/syslog.txt &`

the log will log everything, and can then be analyzed to figure out which app does it and why

**Reply by sideup66** [reply]
Okay, will do! Is there any way to turn on the syslog in this? When I checked other logs I meant dmesg, v3_init.log, wz_post.log, etc. nothing seemed to show there.

**Comment by gtxaspec** [comment]
those should be logging automatically, there should always be some logging in there.  try reformatting the memory card.

**Comment by endertable** [comment]
The following command gives some info on many UDP packets reaching 3 IPs on port 10001 pretty often, along with the occasional NS lookup of www.google.com:

tcpdump -i wlan0 src host 192.x.x.x and udp -Xvvv -nn

where 192.x.x.x is IP of camera.  I believe this is iCamera causing the ruckus on mine, but it may be kvsWebrtcClient as well..

**Reply by sideup66** [reply]
Thats actually the smoking gun. I see it all the time for port 10001 I believe it was too when I see the flood. Didnt realize it was a UDP packet however.... wonder if it gets excessive uring a flood. Be awesome to have a solid trigger to maybe trigger a reboot off of.

**Reply by endertable** [reply]
He included iptables in this mod so you can use that to block that specific outbound port.  I probably won't be messing with it yet because I also use browsers (WebRTC) to view my cams and I don't want to block kvsWebrtcClient from working.  :)

**Reply by sideup66** [reply]
I wonder if a rate limit could solve the problem, flag new and rate limit at maybe 60/min....janky, but it could slow it

**Comment by sideup66** [comment]
Oh I meant there are logs but I meant nothing in the way of a syslog. There was data however. Just nothing useful to investigate this I meant

**Comment by gtxaspec** [comment]
it would be nice to see what iCamera is doing while the flooding is happening, if it's a routine present in an external library we could patch it.

**Comment by gtxaspec** [comment]
or whatever program is causing it.

**Comment by sideup66** [comment]
Is there any way in the libraries maybe we can trick this thing into logging? The wyze firmware seems to kinda be a black box. The one library I see referenced in a file of iCamera in /system/bin is:
/lib/ld-uClibc.so.0 system seems to also have c++ libraries present as well. could we patch one of those maybe?

**Comment by sideup66** [comment]
Just brainstorming about this a bit today to revisit....I am wondering if the problem is triggered by the APP ITSELF losing connection to the camera for whatever reason and not so much the camera itself....does anyone know if the app has any logs we can look at? I hvent had time in the last few days to test but im tempted to setup a pcap to the camera to monitor 10001, then maybe from the app side toggle airplane mode when viewing a stream. Im curious if the connection disruption would trigger this. When I have some time maybe I can perform a simple test.

**Reply by gtxaspec** [reply]
see if this works for you:

https://github.com/gtxaspec/wz_mini_hacks/discussions/94#discussioncomment-2982858

**Reply by sideup66** [reply]
Will give it a shot....problem is to test Id really like to find a way to force the condition to occur. As it sits right now im only speculating its DNS.... not sure what port he saw the traffic on. In our case the flood was on 10001 and NOT 53. When I do my test I can attempt with and without that hosts entry. Like I said the problem is reliably reproducing the bug.....but I think that might be close to the problem....should see if we see traffic over port 10001 to that server. If we do, I think we found it.

**Reply by sideup66** [reply]
Havent noticed any wonkiness since then. Added a pull request to do this in a more automated way that will deal with IP address changes.

**Comment by tavicu** [comment]
It is possible to setup this with the new `rc.local.d` feature?
I don't know for what reason once every two days I need to restart the cameras or frigate will start throwing errors and missing recordings, even if the RTSP is still available (tested with VLC). A restart of the camera fixes it.

**Comment by sideup66** [comment]
hi, yes it is entirely possible to add the command to a script in the rc.d feature.... all youd need to do in there is make a script....say call it 

start_cron.sh
paste in the code below:

#!/bin/sh 
##Verify the network is up before continuing
until ping -c1 www.google.com >/dev/null 2>&1; do :; done
/opt/wz_mini/tmp/.bin/crond -b -c /opt/wz_mini/etc/crontab/


as a matter of fact, im trying to get this integrated in via a pull request here: 
https://github.com/gtxaspec/wz_mini_hacks/pull/148

you might even be able to just download my fork which will have all of that preconfigured (hosts and cron fix):
https://github.com/sideup66/wz_mini_hacks

then you can just set the crontab file to do your reboot as you need.


as for your cameras needing restart, have you confirmed airtime fairness is DISABLED for the 2.4 band? With that on you can get all kinds of weird issues with IoT devices....

**Reply by tavicu** [reply]
Hei. I just checked, Airtime Fairness is already disabled on the router.

**Reply by sideup66** [reply]
Awesome, then yeah if you have the latest build id recommend as you said putting the cron starter in the rc.local.d folder. Id also reccomend to make sure you have the DNS script as it deals with automatically updating the dns entry. The fork I had posted also has the dns script broken out and can be scheduled by cron. I included logic to automatically rebuild the file if it exists vs a full generation if it does not.

**Comment by MomosX** [comment]
Thank you @sideup66 for this guide.
Need a bit of help to get this automatic reboot going - 100% the issue lies with my skillz.

So what i did so far:
- created the directory crontab in /opt/wz_mini/etc 
- created a file named root in /opt/wz_mini/etc/crontab/
- inserted the reboot command in the file root - 45 05 * * * /sbin/reboot . I checked camera time with date command and adjusted accordingly.
- then i inserted the line #!/bin/sh sleep 15 /opt/wz_mini/tmp/.bin/crond -b -c /opt/wz_mini/etc/crontab/ into wz_mini.conf 
![image](https://user-images.githubusercontent.com/61582845/189823403-8d47f8db-5a03-45ec-aa1f-5777f3003360.png)
- reboot
- checked if cron is running with ps ax | grep crond and i get :
![image](https://user-images.githubusercontent.com/61582845/189823599-af16621e-6ab3-40f3-b608-324e718dd643.png)

Config file now looks like : 
![image](https://user-images.githubusercontent.com/61582845/189823753-2b813b4e-62d0-4762-a6c4-54fe5f6c1c01.png)

Is all this correct please? as the camera does not reboot .

Thank you !

**Reply by sideup66** [reply]
Few things here....Cron was integrated into the conf file and defualt startup in https://github.com/gtxaspec/wz_mini_hacks/pull/216

To enable should be simpler now. all you'll need to do is edit the line "ENABLE_CRONTAB" to true in /opt/wz_mini/wz_mini.conf . That will start cron.

Looking at your screenshots posted, cron IS running, however its not reading from /the /etc/crontab, its reading from /etc/cron.

All you need to do is put your command in /opt/wz_mini/etc/cron/root .
Easiest way is to uncomment the line we added which IS a valid reboot command and adjust the time to your needs. Once done you can either manually kill and restart cron yourself or just reboot the camera to force a new crontab (boot time is low, maybe 5-10 seconds if that). 
The cron expression I provided can be copy/pasted and adjusted for time (note the camera's clock is in UTC). If you need help bulding a cron expression the website https://crontab.guru/ is a handy resource :). Let me know if you have any questions.

**Reply by MomosX** [reply]
Hah it was easy.  So it seems the crontab directory is not needed anymore  and the root file is already created in /opt/wz_mini/etc/cron/.
Cron was already enabled in the web GUI, so all i had to do is insert the line  45 05 * * * /sbin/reboot in the existing root file and bam ! it works !!

How can cron be restarted without a full reboot in case i modify the root file ?

Cheers and many thanks about the swith response !

**Reply by sideup66** [reply]
Simplest way to do this without a reboot is to kill the process (i believe #pkill cron may work, otherwise youll just grab the pid with ps ax | grep cron and then use kill $pid). Then you would edit the file, and just restart cron with the command used at startup. An easy tip would be to copy the string from your ps, as that is exact command you need and run that.

**Comment by MomosX** [comment]
pkill cron works. 

All good now:) happy camper :)

Thank you very much !
